### DISCUSSION TITLE
Rust Support

### DISCUSSION THREAD

**Discussion by astrometrics** [apache/netbeans#7684] [Ideas]
Hi, thanks for Netbeans
  I´d like to propose NB to support Rust. In my view it is a good fit and there's a bit of a lack of Rust IDEs, Code being the most common option. It should be a great addition.
Cheers

**Comment by mbien** [comment]
there is some support for rust already in the repo, but it hasn't made it into a release yet.

if you build the `full` cluster config, you can test it:
https://github.com/apache/netbeans?tab=readme-ov-file#building-netbeans

cc @vieiro

**Reply by Chris2011** [reply]
@astrometrics please see this answer, you can already try it out and give feedback and also, if possible you can help here.

**Comment by vieiro** [comment]
You can use `ant -Dcluster.config=rust` to build Rust support.

There're lots of things to do. Maybe we should move this to a plugin for faster development?

**Reply by mbien** [reply]
> Maybe we should move this to a plugin for faster development?

how would this make development faster? Is there a bottleneck somewhere atm?

**Reply by vieiro** [reply]
I'm doing some experiments with Maven plugins and lombok is a great time saver... :-D

**Reply by matthiasblaesing** [reply]
@vieiro where does lombok help? Generating setters and getters? That can to do the IDE for you. I looked at lombok two decades ago (at least that is my feeling), it looked nice, it broke easy. The latter is still true today.

**Reply by neilcsmith-net** [reply]
> Maybe we should move this to a plugin for faster development?

I seem to recall the same argument was used for having it in master?!  How useful is it in its current state?  It's probably better to mark it experimental in the UI and start to include it in the main release if we want to speed up development and contribution?

PS.  If you want to see the fastest veto in history, it'll be me when anyone suggests using Lombok in any of our repos! :rofl:

**Reply by matthiasblaesing** [reply]
I think the real question is: Do we want to try to implement rust support with Java or is integration of rust-analyser the more realistic way? (aka "rustlite")

**Reply by vieiro** [reply]
Well, we can do both. Current `ide/lsp.client` doesn't work well with RustAnalyzer, though, so that will need a change and/or a rewrite.

**Reply by matthiasblaesing** [reply]
What does not work well with `ide/lsp.client`? I have a local branch where I had a prototype. The problem was, that the "java" rust support and the "rust-analyser" support interfered and produced duplicated results (structure trees and such). 

Edit: this was the prototype: https://github.com/matthiasblaesing/netbeans/tree/rust-analyser

Working on that had the consequence, that improvements were intregrated into `lsp.client` (I extracted them as they were general fixes for LSP support).

**Reply by vieiro** [reply]
Last time I tried (a year ago or so) RustAnalyzer complained that it could not be properly initialized (it didn't send the [InitializedNotification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initialized) IIRC.

I think there was a PR that solved the problem, but then it complained about files being opened without being initialized (there was a race condition there, as `ide/lsp.client` uses different threads to talk to the LSP servers).

**Reply by mbien** [reply]
Lombok is its own language by now. Btw the rust modules seem to be still on java 8 language level, if you want records or something simply bump it to 17.

regarding prototyping using maven: I do this often too, since creating lib wrappers is a PITA (esp due to transitive deps), so i do those only when I already know what is needed. (last time i did this was https://github.com/apache/netbeans/issues/7572)

**Comment by astrometrics** [comment]
Hi, good to see some engagement.
Are we in a position now we are able to materialize a plan?

**Comment by astrometrics** [comment]
Hi,
  Chilled question: do you guys think it's possible to start working on rust support now or on a near future? Is it to early, in terms of libraries, etc,  to do so? What would be the main impediments for that?
Cheers
